src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -0,0 +1,731 @@
+
+                    if (bytesRead != 0)
+                    {
+                        buffer.Commit(bytesRead);
+                    }
+                    else if (buffer.ActiveLength == 0)
+                    {
+                        // End of stream.
+                        return (null, 0);
+                    }
+                    else
+                    {
+                        // Our buffer has partial frame data in it but not enough to complete the read: bail out.
+                        throw new Http3ConnectionException(Http3ErrorCode.ClosedCriticalStream);
+                    }
+                }
+
+                buffer.Discard(bytesRead);
+
+                return ((Http3FrameType)frameType, payloadLength);
+            }
+
+            async ValueTask ProcessSettingsFrameAsync(long settingsPayloadLength)
+            {
+                if (settingsPayloadLength > MaximumSettingsPayloadLength)
+                {
+                    throw new HttpRequestException($"SETTINGS frame with payload of length {settingsPayloadLength} is too large; exceeds {MaximumSettingsPayloadLength} bytes.");
+                }
+
+                while (settingsPayloadLength != 0)
+                {
+                    long settingId, settingValue;
+                    int bytesRead;
+
+                    while (!Http3Frame.TryReadIntegerPair(buffer.ActiveSpan, out settingId, out settingValue, out bytesRead))
+                    {
+                        buffer.EnsureAvailableSpace(1);
+                        bytesRead = await stream.ReadAsync(buffer.AvailableMemory, CancellationToken.None).ConfigureAwait(false);
+
+                        if (bytesRead != 0)
+                        {
+                            buffer.Commit(bytesRead);
+                        }
+                        else
+                        {
+                            // Our buffer has partial frame data in it but not enough to complete the read: bail out.
+                            throw new Http3ConnectionException(Http3ErrorCode.ClosedCriticalStream);
+                        }
+                    }
+
+                    settingsPayloadLength -= bytesRead;
+
+                    // Only support this single setting. Skip others.
+                    if (settingId == (long)Http3SettingType.MaxHeaderListSize)
+                    {
+                        _maximumHeadersLength = (int)Math.Min(settingValue, int.MaxValue);
+                    }
+                }
+            }
+
+            async ValueTask ProcessGoAwayFameAsync(long goawayPayloadLength)
+            {
+                long lastStreamId;
+                int bytesRead;
+
+                while (!VariableLengthIntegerHelper.TryRead(buffer.AvailableSpan, out lastStreamId, out bytesRead))
+                {
+                    buffer.EnsureAvailableSpace(1);
```

Comment:
This seems inefficient. This is only growing the buffer by one each time, which causes TryRead and ReadAsync to be called multiple times if we are reading a large integer.

Reply:
It looks like ArrayBuffer [grows by doubling](https://github.com/dotnet/runtime/blob/1eb74d072e930a87955ed856a27866535ff165b7/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ArrayBuffer.cs#L103-L109) even if only 1 byte is needed

Reply:
Okay not as bad as I initially thought. Maybe something should be renamed to make it more clear that we are increasing the space by more than just 1 (at least double).

I still think this pattern of growing the buffer is prone to smaller reads though.

Reply:
Current strategy is to: read in just enough to parse headers in one I/O, and then let user control read pattern through Content.

Once we get implementation finished, we'll want to tune this. It's not clear what buffer size will be optimal just yet, but it will be trivial to change when we're ready.